File: src/recording/typer.ts
/* Pure typing primitives shared by every sim.

   A sim turns its script into a flat list of TypeSteps; the typewriter hook runs
   them. Body text may contain the legacy typo markup [[wrong|right]] — type the
   wrong spelling, pause, backspace it, type the correction — which we expand into
   low-level keystroke actions here. */

export type TypeStep =
  | { kind: 'type'; text: string; onUpdate: (partial: string) => void; cue?: number }
  | { kind: 'reveal'; fn: () => void; delay?: number }
  | { kind: 'pause'; ms: number };

export type KeyAction = { t: 'char'; ch: string } | { t: 'back' } | { t: 'wait'; ms: number };

export const sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

/** Expand a body string (with optional [[wrong|right]] typos) into keystrokes. */
export function tokenize(text: string): KeyAction[] {
  const out: KeyAction[] = [];
  const pushChars = (s: string) => { for (const ch of Array.from(s)) out.push({ t: 'char', ch }); };
  const re = /\[\[([^\]|]*)\|([^\]]*)\]\]/g;
  let last = 0;
  let m: RegExpExecArray | null;
  while ((m = re.exec(text))) {
    pushChars(text.slice(last, m.index));
    const wrong = m[1];
    const right = m[2];
    pushChars(wrong);
    out.push({ t: 'wait', ms: 240 + Math.random() * 160 });
    for (let i = 0; i < Array.from(wrong).length; i++) out.push({ t: 'back' });
    out.push({ t: 'wait', ms: 110 + Math.random() * 90 });
    pushChars(right);
    last = re.lastIndex;
  }
  pushChars(text.slice(last));
  return out;
}

/** Strip the typo markup to get the final, settled text (for instant fills). */
export function settledText(text: string): string {
  return text.replace(/\[\[([^\]|]*)\|([^\]]*)\]\]/g, (_, _w, right) => right);
}

export interface TimingOpts { jitter: number; thinkPauses: number }

/** Per-character delay (ms). jitter scales variance; thinkPauses scales the
    occasional hesitation. Spaces/punctuation breathe; letters are quick. */
export function charDelay(ch: string, o: TimingOpts): number {
  const j = 0.35 + o.jitter; // keep a little variance even at 0
  if (ch === '\n') return 170 + Math.random() * 150 * j;
  if (ch === ' ') return 52 + Math.random() * 55 * j;
  if (',.;:!?—'.includes(ch)) return 115 + Math.random() * 130 * j;
  const base = 40 + Math.random() * 65 * j;
  const hitchProb = 0.1 * o.thinkPauses;
  return Math.random() < hitchProb ? base + 150 + Math.random() * 260 * o.thinkPauses : base;
}

/** A plausible wrong neighbor key for the auto-typo effect. */
const NEIGHBORS: Record<string, string> = {
  a: 's', s: 'd', d: 'f', f: 'g', g: 'h', h: 'j', j: 'k', k: 'l', l: 'k',
  q: 'w', w: 'e', e: 'r', r: 't', t: 'y', y: 'u', u: 'i', i: 'o', o: 'p', p: 'o',
  z: 'x', x: 'c', c: 'v', v: 'b', b: 'n', n: 'm', m: 'n',
};
export function fumbleFor(ch: string): string | null {
  const lower = ch.toLowerCase();
  const n = NEIGHBORS[lower];
  if (!n) return null;
  return ch === lower ? n : n.toUpperCase();
}

File: app/src/main/java/com/example/recorder/engine/Typer.kt
package com.example.recorder.engine

import kotlin.random.Random

/**
 * Pure typing primitives shared by every sim. Port of src/recording/typer.ts.
 *
 * A sim turns its script into a flat list of [TypeStep]s; [TypeEngine] runs them.
 * Body text may contain the typo markup [[wrong|right]] — type the wrong spelling,
 * pause, backspace it, type the correction — expanded into keystrokes by [tokenize].
 */

sealed interface TypeStep {
    /** Type [text] character-by-character, pushing each partial to [onUpdate]. */
    data class Type(
        val text: String,
        val onUpdate: (String) -> Unit,
        val cue: Float? = null,
    ) : TypeStep

    /** Run [fn] (reveal a card, reset state), then optionally wait [delay] ms. */
    data class Reveal(val fn: () -> Unit, val delay: Int? = null) : TypeStep

    /** Idle for [ms] milliseconds. */
    data class Pause(val ms: Int) : TypeStep
}

sealed interface KeyAction {
    data class Char(val ch: String) : KeyAction
    data object Back : KeyAction
    data class Wait(val ms: Int) : KeyAction
}

private val TYPO_RE = Regex("""\[\[([^\]|]*)\|([^\]]*)]]""")

/** Expand a body string (with optional [[wrong|right]] typos) into keystrokes. */
fun tokenize(text: String): List<KeyAction> {
    val out = ArrayList<KeyAction>()
    fun pushChars(s: String) {
        // iterate by code points so emoji/surrogate pairs stay intact
        var i = 0
        while (i < s.length) {
            val cp = s.codePointAt(i)
            val n = Character.charCount(cp)
            out.add(KeyAction.Char(s.substring(i, i + n)))
            i += n
        }
    }
    var last = 0
    for (m in TYPO_RE.findAll(text)) {
        pushChars(text.substring(last, m.range.first))
        val wrong = m.groupValues[1]
        val right = m.groupValues[2]
        pushChars(wrong)
        out.add(KeyAction.Wait((240 + Random.nextDouble() * 160).toInt()))
        val wrongLen = wrong.codePointCount(0, wrong.length)
        repeat(wrongLen) { out.add(KeyAction.Back) }
        out.add(KeyAction.Wait((110 + Random.nextDouble() * 90).toInt()))
        pushChars(right)
        last = m.range.last + 1
    }
    pushChars(text.substring(last))
    return out
}

/** Strip the typo markup to get the final, settled text (for instant fills/preview). */
fun settledText(text: String): String =
    TYPO_RE.replace(text) { it.groupValues[2] }

data class TimingOpts(val jitter: Float, val thinkPauses: Float)

/**
 * Per-character delay (ms). jitter scales variance; thinkPauses scales the
 * occasional hesitation. Spaces/punctuation breathe; letters are quick.
 */
fun charDelay(ch: String, o: TimingOpts): Double {
    val j = 0.35 + o.jitter // keep a little variance even at 0
    if (ch == "\n") return 170 + Random.nextDouble() * 150 * j
    if (ch == " ") return 52 + Random.nextDouble() * 55 * j
    if (ch.length == 1 && ",.;:!?—".contains(ch)) return 115 + Random.nextDouble() * 130 * j
    val base = 40 + Random.nextDouble() * 65 * j
    val hitchProb = 0.1 * o.thinkPauses
    return if (Random.nextDouble() < hitchProb)
        base + 150 + Random.nextDouble() * 260 * o.thinkPauses
    else base
}

/** A plausible wrong neighbor key for the auto-typo effect. */
private val NEIGHBORS: Map<Char, Char> = mapOf(
    'a' to 's', 's' to 'd', 'd' to 'f', 'f' to 'g', 'g' to 'h', 'h' to 'j', 'j' to 'k', 'k' to 'l', 'l' to 'k',
    'q' to 'w', 'w' to 'e', 'e' to 'r', 'r' to 't', 't' to 'y', 'y' to 'u', 'u' to 'i', 'i' to 'o', 'o' to 'p', 'p' to 'o',
    'z' to 'x', 'x' to 'c', 'c' to 'v', 'v' to 'b', 'b' to 'n', 'n' to 'm', 'm' to 'n',
)

fun fumbleFor(ch: String): String? {
    if (ch.length != 1) return null
    val c = ch[0]
    val lower = c.lowercaseChar()
    val n = NEIGHBORS[lower] ?: return null
    return if (c == lower) n.toString() else n.uppercaseChar().toString()
}

/** A plausible mistyped version of a whole word (one neighbor-key swap). */
fun fumbleWord(word: String): String {
    if (word.length < 2) return word
    val i = Random.nextInt(word.length)
    val rep = fumbleFor(word[i].toString()) ?: return word
    return word.substring(0, i) + rep + word.substring(i + 1)
}

/** Split text into ordered (segment, isWord) runs — words vs whitespace runs. */
fun splitWords(t: String): List<Pair<String, Boolean>> {
    if (t.isEmpty()) return emptyList()
    val res = ArrayList<Pair<String, Boolean>>()
    var i = 0
    while (i < t.length) {
        val ws = t[i].isWhitespace()
        val start = i
        while (i < t.length && t[i].isWhitespace() == ws) i++
        res.add(t.substring(start, i) to !ws)
    }
    return res
}
